runtime/jvmti/suspendhelper.cpp

```diff
@@ -45,38 +45,48 @@ suspendThread(J9VMThread *currentThread, jthread thread, BOOLEAN allowNull, BOOL
-					rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
+				if (OMR_ARE_ANY_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND)) {
+					J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedInternalOffset, 1);
 				} else {
-					if (currentThread == targetThread) {
-						*currentThreadSuspended = TRUE;
+					if (OMR_ARE_ANY_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_STOPPED)) {
+						/* Do not suspend dead threads. Mirrors SUN behaviour. */
+						rc = JVMTI_ERROR_THREAD_NOT_ALIVE;
 					} else {
-						vm->internalVMFunctions->internalExitVMToJNI(currentThread);
-						omrthread_monitor_enter(targetThread->publicFlagsMutex);
-						VM_VMAccess::setHaltFlagForVMAccessRelease(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
-						if (VM_VMAccess::mustWaitForVMAccessRelease(targetThread)) {
-							while (J9_ARE_ALL_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND | J9_PUBLIC_FLAGS_VM_ACCESS)) {
-								omrthread_monitor_wait(targetThread->publicFlagsMutex);
+						if (currentThread == targetThread) {
+#if JAVA_SPEC_VERSION >= 19
+							J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedInternalOffset, 1);
+#endif /* JAVA_SPEC_VERSION >= 19 */
+							*currentThreadSuspended = TRUE;
+						} else {
+							vm->internalVMFunctions->internalExitVMToJNI(currentThread);
+#if JAVA_SPEC_VERSION >= 19
+							J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedInternalOffset, 1);
+#endif /* JAVA_SPEC_VERSION >= 19 */
+							omrthread_monitor_enter(targetThread->publicFlagsMutex);
+							VM_VMAccess::setHaltFlagForVMAccessRelease(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
+							if (VM_VMAccess::mustWaitForVMAccessRelease(targetThread)) {
+								while (J9_ARE_ALL_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND | J9_PUBLIC_FLAGS_VM_ACCESS)) {
+									omrthread_monitor_wait(targetThread->publicFlagsMutex);
+								}
 							}
+							omrthread_monitor_exit(targetThread->publicFlagsMutex);
+							vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
 						}
-						omrthread_monitor_exit(targetThread->publicFlagsMutex);
-						vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
+						Trc_JVMTI_threadSuspended(targetThread);
 					}
-					Trc_JVMTI_threadSuspended(targetThread);
 				}
 			}
 		}
 #if JAVA_SPEC_VERSION >= 19
 		else {
 			/* targetThread is NULL only for virtual threads as per the assertion in getVMThread. */
-			if (0 != J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedByJVMTIOffset)) {
+			if (0 != J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset)) {
 				rc = JVMTI_ERROR_THREAD_SUSPENDED;
 			} else {
-				J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedByJVMTIOffset, 1);
+				J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedInternalOffset, 1);
 			}
```

Comment:
This logic is common between carrier and virtual threads. The else block can be removed so that it can be shared. Also, `threadObject` needs to be refetched because VM access is released and reacquired while setting the suspend flag; GC can move objects during this gap.

```
threadObject = (NULL == thread) ? currentThread->threadObject : J9_JNI_UNWRAP_REFERENCE(thread);
```

Reply:
This comment has not been addressed. The else block needs to be removed and its code should be shared.
```
		if ((NULL != targetThread) && (threadObject == targetThread->threadObject))
#endif /* JAVA_SPEC_VERSION >= 19 */
		{
			...
		}
#if JAVA_SPEC_VERSION >= 19
		if (0 != J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset)) {
			rc = JVMTI_ERROR_THREAD_SUSPENDED;
		} else {
			J9OBJECT_U32_STORE(currentThread, threadObject, vm->isSuspendedInternalOffset, 1);
		}
#endif /* JAVA_SPEC_VERSION >= 19 */
```

Reply:
else block is removed to handle the common code.